clvm-derive/src/to_clvm.rs

```diff
@@ -1,53 +1,230 @@
+use syn::{
+    parse_quote, spanned::Spanned, Data, DeriveInput, Expr, Fields, FieldsNamed, FieldsUnnamed,
+    GenericParam, Index,
+};
 
-use crate::helpers::{add_trait_bounds, parse_args, Repr};
+use crate::{
+    helpers::{add_trait_bounds, parse_clvm_attr, Repr},
+    macros::{repr_macros, Macros},
+};
 
-pub fn to_clvm(mut ast: DeriveInput) -> TokenStream {
-    let args = parse_args(&ast.attrs);
-    let crate_name = quote!(clvm_traits);
+#[derive(Default)]
+struct FieldInfo {
+    field_names: Vec<Ident>,
+    field_accessors: Vec<TokenStream>,
+    initializer: TokenStream,
+}
+
+struct VariantInfo {
+    name: Ident,
+    value: Expr,
+    field_info: FieldInfo,
+    macros: Macros,
+}
 
-    let field_names: Vec<TokenStream>;
+pub fn to_clvm(ast: DeriveInput) -> TokenStream {
+    let clvm_attr = parse_clvm_attr(&ast.attrs);
+    let crate_name = quote!(clvm_traits);
 
     match &ast.data {
-        Data::Struct(data_struct) => match &data_struct.fields {
-            Fields::Named(fields) => {
-                let fields = &fields.named;
-                field_names = fields
-                    .iter()
-                    .map(|field| {
-                        let ident = field.ident.clone().unwrap();
-                        quote!(#ident)
-                    })
-                    .collect();
+        Data::Struct(data_struct) => {
+            if clvm_attr.untagged {
+                panic!("cannot use `untagged` on a struct");
             }
-            Fields::Unnamed(fields) => {
-                let fields = &fields.unnamed;
-                field_names = fields
-                    .iter()
-                    .enumerate()
-                    .map(|(i, _)| {
-                        let index = Index::from(i);
-                        quote!(#index)
-                    })
-                    .collect();
+            let macros = repr_macros(&crate_name, clvm_attr.expect_repr());
+            let field_info = fields(&data_struct.fields);
+            impl_for_struct(&crate_name, &ast, &macros, &field_info)
+        }
+        Data::Enum(data_enum) => {
+            if !clvm_attr.untagged && clvm_attr.repr == Some(Repr::Curry) {
+                panic!("cannot use `curry` on a tagged enum");
+            }
+
+            let mut next_value: Expr = parse_quote!(0);
+            let mut variants = Vec::new();
+
+            for variant in data_enum.variants.iter() {
+                let field_info = fields(&variant.fields);
+                let variant_clvm_attr = parse_clvm_attr(&variant.attrs);
+
+                if variant_clvm_attr.untagged {
+                    panic!("cannot use `untagged` on an enum variant");
+                }
+
+                let repr = variant_clvm_attr
+                    .repr
+                    .unwrap_or_else(|| clvm_attr.expect_repr());
+                if !clvm_attr.untagged && repr == Repr::Curry {
+                    panic!("cannot use `curry` on a tagged enum variant");
```

Comment:
why is this combination of attributes disallowed? (maybe the error message should say)

Reply:
Explained it in the error - Unlike other representations, each argument in a curried parameter list is wrapped so it doesn't work with the two step parsing method I'm using.

In the future, I will write utilities to simplify parsing list/tuple/curry/similar structures, and remove this limitation.